File: frontend/src/api/ai.ts
import request from './request'

export interface ProviderMeta {
  provider_key: string
  display_name: string
  default_base_url: string | null
  suggested_models: string[]
}

export interface AiCredential {
  id: number
  provider: string
  model_name: string
  base_url: string | null
  enabled: boolean
  has_api_key: boolean
  created_at: string
}

export interface AiStrategy {
  id: number
  name: string
  strategy_text: string
  active: boolean
  credential_id: number | null
  created_at: string
  updated_at: string
}

export const aiApi = {
  providers: () => request.get<unknown, ProviderMeta[]>('/ai/providers'),

  listCredentials: () => request.get<unknown, AiCredential[]>('/ai/credentials'),
  createCredential: (body: {
    provider: string; model_name: string; api_key: string; base_url?: string | null; enabled?: boolean
  }) => request.post<unknown, AiCredential>('/ai/credentials', body),
  patchCredential: (id: number, body: Partial<{ model_name: string; api_key: string; base_url: string | null; enabled: boolean }>) =>
    request.patch<unknown, AiCredential>(`/ai/credentials/${id}`, body),
  deleteCredential: (id: number) =>
    request.delete<unknown, { deleted: number }>(`/ai/credentials/${id}`),

  listStrategies: () => request.get<unknown, AiStrategy[]>('/ai/strategies'),
  createStrategy: (body: Partial<AiStrategy>) =>
    request.post<unknown, AiStrategy>('/ai/strategies', body),
  patchStrategy: (id: number, body: Partial<AiStrategy>) =>
    request.patch<unknown, AiStrategy>(`/ai/strategies/${id}`, body),
  deleteStrategy: (id: number) =>
    request.delete<unknown, { deleted: number }>(`/ai/strategies/${id}`),

  preview: (strategy_text: string) =>
    request.post<unknown, { prompt: string }>('/ai/strategies/preview', { strategy_text }),

  test: (body: { credential_id: number; strategy_id?: number | null; sample: Record<string, unknown> }) =>
    request.post<unknown, { category: string | null; confidence: string | null; raw: string }>('/ai/test', body),
}

File: backend/app/schemas/ai.py
from datetime import datetime

from pydantic import BaseModel, Field


class AiCredentialIn(BaseModel):
    provider: str = Field(min_length=1, max_length=32)
    model_name: str = Field(min_length=1, max_length=64)
    api_key: str = Field(min_length=1)
    base_url: str | None = Field(default=None, max_length=255)
    enabled: bool = True


class AiCredentialPatchIn(BaseModel):
    model_name: str | None = Field(default=None, max_length=64)
    api_key: str | None = None
    base_url: str | None = Field(default=None, max_length=255)
    enabled: bool | None = None


class AiCredentialOut(BaseModel):
    id: int
    provider: str
    model_name: str
    base_url: str | None
    enabled: bool
    has_api_key: bool = True
    created_at: datetime

    model_config = {"from_attributes": True}


class AiStrategyIn(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    strategy_text: str = ""
    active: bool = False
    credential_id: int | None = None


class AiStrategyPatchIn(BaseModel):
    name: str | None = Field(default=None, max_length=64)
    strategy_text: str | None = None
    active: bool | None = None
    credential_id: int | None = None


class AiStrategyOut(BaseModel):
    id: int
    name: str
    strategy_text: str
    active: bool
    credential_id: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PreviewIn(BaseModel):
    strategy_text: str = ""


class PreviewOut(BaseModel):
    prompt: str


class TestIn(BaseModel):
    credential_id: int
    strategy_id: int | None = None
    sample: dict


class TestOut(BaseModel):
    category: str | None
    confidence: str | None
    raw: str


class ProviderMetaOut(BaseModel):
    provider_key: str
    display_name: str
    default_base_url: str | None
    suggested_models: list[str]
